testing_and_setup/compass/manage_regression_suite.py

```diff
@@ -193,19 +196,282 @@ def process_test_clean(test_tag, work_dir):  # {{{
+    local_parallel_code += "  print('more processors are used than allocated')\n"
+    local_parallel_code += "  quit()\n"
+    local_parallel_code += "start_time = time.time()\n"
+    local_parallel_code += "def stream_queue(command, number_of_procs, procs):\n"
+    local_parallel_code += "  if command == []:\n"
+    local_parallel_code += "    return \n"
+    local_parallel_code += "  base = '{}/'\n".format(work_dir)
+    local_parallel_code += "  Queue_running = []\n"
+    local_parallel_code += "  index = 0\n"
+    local_parallel_code += "  continue_add = True\n"
+    local_parallel_code += "  Done = False\n"
+    local_parallel_code += "  while True:\n"
+    local_parallel_code += "     if continue_add and not Done:\n"
+    local_parallel_code += "      if number_of_procs >= procs[index]:\n"
+    local_parallel_code += "        try:\n"
+    local_parallel_code += "          if index < len(command):\n"
+    local_parallel_code += "#            print('-----IN ADDING PHASE-----')\n"
+    local_parallel_code += "#            print('we have {} number_of_procs, current task uses {} Adding to queue'.format(number_of_procs, procs[index]))\n"
+    local_parallel_code += "            case_output = open('case_outputs/'+command[index][0].replace('/', '_'), 'w+')\n"
+    local_parallel_code += "#            print('processing command: {}'.format(command[index][1]))\n"
+    local_parallel_code += "            os.chdir(command[index][0])\n"
+    local_parallel_code += "            command[index][1][2] = command[index][1][2] +'/'+data_dic['run_file_names'][index]\n"
+    local_parallel_code += "            open_proc = subprocess.Popen(command[index][1], stdout=case_output, stderr=case_output)\n"
+    local_parallel_code += "            os.chdir(base)\n"
+    local_parallel_code += "            Queue_running.append([open_proc, procs[index],command[index][1]]) \n"
+    local_parallel_code += "            number_of_procs = number_of_procs - procs[index]\n"
+    local_parallel_code += "#            print('       New number_of_procs: {}'.format(number_of_procs)) \n"
+    local_parallel_code += "        except subprocess.CalledProcessError: \n"
+    local_parallel_code += "          print('error for : {}'.format(str(command[index][1]))) \n"
+    local_parallel_code += "      elif number_of_procs < procs[index]: \n"
+    local_parallel_code += "#        print('NOT ENOUGH WAIT FOR PROCESSES') \n"
+    local_parallel_code += "        continue_add = False \n"
+    local_parallel_code += "      index = index + 1 \n"
+    local_parallel_code += "      if index > len(procs) -1: \n"
+    local_parallel_code += "#        print('No more to add moving to processing phase') \n"
+    local_parallel_code += "        continue_add = False \n"
+    local_parallel_code += "     elif not continue_add and not Done: \n"
+    local_parallel_code += "       for background_process in Queue_running: \n"
+    local_parallel_code += "         background_process[0].wait() \n"
```

Comment:
@jamilgafur, let me know if I've misunderstood but doesn't this wait for the first process in the queue, rather than the one that finishes first?  Wouldn't this be really inefficient if the first process happens to be one that takes a really long time?  Don't you want to go through all the running processes instead and see which (if any) is done, go to sleep for a little while, and check again?

Reply:
I think the right method to call might be `.poll()`: 
https://docs.python.org/2/library/subprocess.html#subprocess.Popen.poll
https://stackoverflow.com/a/12058609/7728169

Reply:
there is a timeout argument in the wait function that I utilized